src/build/check/mod.rs

```diff
@@ -0,0 +1,262 @@
+///         Ok(Self(*input))
+///     }
+/// }
+///
+/// let n = SmallNumber::parse(&8);
+///
+/// assert!(n.is_ok());
+/// ```
+pub trait Parseable<ParserInput>: Sized {
+    fn parse(input: &ParserInput) -> Result<Self, failure::Error>;
+}
+
+/// If a struct is Lintable, that means it's able to check itself according to some
+/// criteria. It can be linted -- it's lintable!
+/// ```
+/// trait Lintable {
+///     fn lint(&self) -> Result<(), failure::Error>;
+/// }
+///
+/// struct SmallNumber(u8);
+///
+/// impl Lintable for SmallNumber {
+///     fn lint(&self) -> Result<(), failure::Error> {
+///         if self.0 > 4 {
+///             Err(failure::err_msg("Number is too big!"))
+///         } else {
+///             Ok(())
+///         }
+///     }
+/// }
+///
+/// let three = SmallNumber(3);
+/// let seven = SmallNumber(7);
+///
+/// assert!(three.lint().is_ok());
+/// assert!(seven.lint().is_err());
+/// ```
+pub trait Lintable {
+    fn lint(&self) -> Result<(), failure::Error>;
+}
+
+/// If a struct is both Parseable and Lintable, then it may also implement
+/// Validate, which should serve as a one-stop-shop to go from un-parsed
+/// input to linted output. A default implmenetation is provided which
+/// simply calls `Self::parse(&input).lint()`
+/// ```
+/// # trait Parseable<ParserInput>: Sized {
+/// #    fn parse(input: &ParserInput) -> Result<Self, failure::Error>;
+/// # }
+/// #
+/// # trait Lintable {
+/// #     fn lint(&self) -> Result<(), failure::Error>;
+/// # }
+/// #
+/// pub trait Validate<ParserInput>:
+///     Lintable + Parseable<ParserInput>
+/// {
+///         fn validate(input: ParserInput) -> Result<(), failure::Error> {
+///             let t = Self::parse(&input)?;
+///             t.lint()
+///     }
+/// }
+///
+/// struct SmallNumber(u8);
+///
+/// impl Parseable<u8> for SmallNumber {
+///     fn parse(input: &u8) -> Result<Self, failure::Error> {
+///         Ok(Self(*input))
+///     }
+/// }
+///
+/// impl Lintable for SmallNumber {
+///     fn lint(&self) -> Result<(), failure::Error> {
+///         if self.0 > 4 {
+///             Err(failure::err_msg("Number is too big!"))
+///         } else {
+///             Ok(())
+///         }
+///     }
+/// }
+///
+/// impl Validate<u8> for SmallNumber {};
+///
+/// assert!(SmallNumber::validate(3).is_ok());
+/// assert!(SmallNumber::validate(6).is_err());
+/// ```
+pub trait Validate<ParserInput>: Lintable + Parseable<ParserInput> {
+    fn validate(input: ParserInput) -> Result<(), failure::Error> {
+        let t = Self::parse(&input)?;
+        t.lint()
+    }
+}
+
+/// Represents the output of any bundler tool.
+///
+/// I'm not 100% on the format of this struct because I don't
+/// have access to the durable objects beta
+/// but it seems as though the format is basically:
+/// * one `.mjs` file that serves as the entrypoint to the worker,
+/// * any number other arbitrary files that can be imported into the worker.
```

Comment:
This is true, but there is a subtle nuance: what is the content-type of each file?

You might want to import an HTML file as an `ArrayBuffer`, thus using `application/octet-stream` in the form data upload; or `text/plain` which would import it as a `String`.

From what I gather, there are 3 potential options:
1. Always upload non-JS and non-WASM files as `application/octet-stream` and worry about this nuance later
2. Have users define the content-type of their imports somewhere in the `wrangler.toml`
3. Adopt the tc39 proposal for [import-assertions](https://github.com/tc39/proposal-import-assertions/), which `swc` now [supports](https://github.com/swc-project/swc/pull/1100)

Moreover, if we decide to adopt the proposal, there are 2 additional considerations:
1. Since V8 does not support it, each assertion would need to be "removed" prior to upload?
2. What will each assertion type be called? "arraybuffer"? "text"?

Coincidentally, our very own @xtuc is championing this proposal, so I think he would have some great ideas.

Reply:
Oh hmm interesting!

I was able to find the [tracking issue](https://bugs.chromium.org/p/v8/issues/detail?id=10958) for V8, and it seems like they're working on it at the moment. I don't know what the typical timeline is with V8 from "started working" to "shipped", but I'd assume it's...a couple months? Is that overly optimistic? I'd love to hear some more input *and* hopefully by the time this PR lands it'll become a non-issue.
 
> 1. Always upload non-JS and non-WASM files as `application/octet-stream` and worry about this nuance later

I choose the option that involves the least work!

> 2. Have users define the content-type of their imports somewhere in the `wrangler.toml`

Speaking from personal experience, the less research and configuring I have to do, the better. Part of the appeal of wrangler (and Workers) is that it "just works", so I'm kind of opposed to this idea but not strongly. Just kind of wary of adding more configuration that for most folks will likely be unnecessary.

> 3. Adopt the tc39 proposal for [import-assertions](https://github.com/tc39/proposal-import-assertions/), which `swc` now [supports](https://github.com/swc-project/swc/pull/1100)

I'm hesitant to start parsing things that don't map directly to V8 because then we might start to get into more opinionated parsing, like you said, which could lead to unexpected behavior somewhere down the line.

In my opinion the optimal approach for getting this done before V8 adopts import-assertions would be to traverse the AST and see what kinds of methods are called on each import, and whether those are `ArrayBuffer` type things or `String` type things, but that seems...overkill?

I think probably the easiest thing to do is to use `application/octet-stream` and document that that's what we've selected, and that if you want to manipulate your import as a `String` you should do some conversion first. [There's a million StackOverflow questions about that exact thing](https://stackoverflow.com/search?q=%5Bjavascript%5D+arraybuffer+string) so I figure that won't be *too* much of a hassle.